modules/promotion/src/PromotionOfferManager.php

```diff
@@ -0,0 +1,91 @@
+use Drupal\Core\Plugin\Context\ContextDefinition;
+use Drupal\Core\Plugin\DefaultPluginManager;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+
+/**
+ * Provides the Offer plugin manager.
+ */
+class PromotionOfferManager extends DefaultPluginManager implements CategorizingPluginManagerInterface {
+
+  use CategorizingPluginManagerTrait;
+
+  /**
+   * Constructor for OfferManager objects.
+   *
+   * @param \Traversable $namespaces
+   *   An object that implements \Traversable which contains the root paths
+   *   keyed by the corresponding namespace to look for plugin implementations.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   Cache backend instance to use.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler to invoke the alter hook with.
+   */
+  public function __construct(\Traversable $namespaces, CacheBackendInterface $cache_backend, ModuleHandlerInterface $module_handler) {
+    parent::__construct('Plugin/Commerce/PromotionOffer', $namespaces, $module_handler, 'Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\PromotionOfferInterface', 'Drupal\commerce_promotion\Annotation\PromotionOffer');
+
+    $this->alterInfo('commerce_promotion_offer_info');
+    $this->setCacheBackend($cache_backend, 'commerce_promotion_offer_plugins');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function createInstance($plugin_id, array $configuration = []) {
+    $plugin = $this->getFactory()->createInstance($plugin_id, $configuration);
+
+    // If we receive any context values via config set it into the plugin.
+    if (!empty($configuration['context'])) {
+      foreach ($configuration['context'] as $name => $context) {
+        $plugin->setContextValue($name, $context);
+      }
+    }
+
+    return $plugin;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function processDefinition(&$definition, $plugin_id) {
+    parent::processDefinition($definition, $plugin_id);
+
+    foreach (['id', 'label', 'target'] as $required_property) {
+      if (empty($definition[$required_property])) {
+        throw new PluginException(sprintf('The promotion offer %s must define the %s property.', $plugin_id, $required_property));
+      }
+    }
+
+    $target = $definition['target'];
+    if (!\Drupal::entityTypeManager()->getDefinition($target)) {
+      throw new PluginException(sprintf('The promotion offer %s must reference a valid entity type, %s given.', $plugin_id, $target));
+    }
+
+    // If the plugin did not specify a category, process that.
+    if (empty($definition['category'])) {
+      if ($target == PromotionOfferInterface::ORDER) {
+        $definition['category'] = $this->t('Order');
+      }
+      elseif ($target == PromotionOfferInterface::LINE_ITEM) {
+        $definition['category'] = $this->t('Product');
+      }
+      else {
+        $definition['category'] = $this->t('Other');
```

Comment:
Shipment? I had no idea what to do here because we can't throw an exception because what if people add promotions targeting other entities. Maybe I just need to load the entity definition and get the entity's label.